python/ray/serve/tests/test_fastapi.py

```diff
@@ -768,6 +768,259 @@ def func(a: int):
+        response.headers["X-Custom-Middleware"] = "fake-middleware"
+        return response
+
+    # custom exception handler
+    @app.exception_handler(ValueError)
+    async def custom_exception_handler(request: Request, exc: ValueError):
+        return JSONResponse(status_code=500, content={"error": "fake-error"})
+
+    return app
+
+
+def test_ingress_with_fastapi_routes_outside_deployment(serve_instance):
+    app = fastapi_builder()
+
+    @serve.deployment
+    @serve.ingress(app)
+    class ASGIIngress:
+        @app.get("/class_route")
+        def class_route(self):
+            return "hello class route"
+
+    serve.run(ASGIIngress.bind())
+    assert requests.get("http://localhost:8000/").json() == "hello"
+    assert requests.get("http://localhost:8000/f2").json() == "hello f2"
+    assert (
+        requests.get("http://localhost:8000/class_route").json() == "hello class route"
+    )
+    assert requests.get("http://localhost:8000/error").status_code == 500
+    assert requests.get("http://localhost:8000/error").json() == {"error": "fake-error"}
+
+    # get the docs path from the controller
+    docs_path = ray.get(serve_instance._controller.get_docs_path.remote("default"))
+    assert docs_path == "/custom-docs"
+
+
+def test_ingress_with_fastapi_with_no_deployment_class(serve_instance):
+    app = fastapi_builder()
+
+    ingress_deployment = serve.deployment(serve.ingress(app)())
+    assert ingress_deployment.name == "ASGIIngressDeployment"
+    serve.run(ingress_deployment.bind())
+    assert requests.get("http://localhost:8000/").json() == "hello"
+    assert requests.get("http://localhost:8000/f2").json() == "hello f2"
+    assert requests.get("http://localhost:8000/error").status_code == 500
+    assert requests.get("http://localhost:8000/error").json() == {"error": "fake-error"}
+
+    # get the docs path from the controller
+    docs_path = ray.get(serve_instance._controller.get_docs_path.remote("default"))
+    assert docs_path == "/custom-docs"
+
+
+def test_ingress_with_fastapi_builder_function(serve_instance):
+    ingress_deployment = serve.deployment(serve.ingress(fastapi_builder)())
+    serve.run(ingress_deployment.bind())
+
+    resp = requests.get("http://localhost:8000/")
+    assert resp.json() == "hello"
+    assert resp.headers["X-Custom-Middleware"] == "fake-middleware"
+
+    resp = requests.get("http://localhost:8000/f2")
+    assert resp.json() == "hello f2"
+    assert resp.headers["X-Custom-Middleware"] == "fake-middleware"
+
+    resp = requests.get("http://localhost:8000/error")
+    assert resp.status_code == 500
+    assert resp.json() == {"error": "fake-error"}
+
+    docs_path = ray.get(serve_instance._controller.get_docs_path.remote("default"))
+    assert docs_path == "/custom-docs"
+
+
+def test_ingress_with_fastapi_builder_with_deployment_class(serve_instance):
+    @serve.deployment
+    @serve.ingress(fastapi_builder)
+    class ASGIIngress:
```

Comment:
from the user's perspective, is there benefits to to combining a fastapi builder function with a deployment class? can the fastapi routes access things on the deployment class?

Reply:
>from the user's perspective, is there benefits to combining a fastapi builder function with a deployment class?
could you ask this question in a different way, I didn't fully understand the Q here.

>can the fastapi routes access things on the deployment class?
When using builder function with routes outside the deployment class, the routes cannot natively access the deployment class (ingress or others), via `self`, how it was possible before. They would need to first get the handle using `serve.get_deployment_handle` api.

Reply:
to put it more bluntly is there a point to passing in a builder function to `serve.ingress`, but also have it decorate a non-null class? or is the api intended to be either one or the other.

Reply:
i cannot think of a use case where user needs the ingress deployment class when using the builder function other than specifying dependent deployments to build out the topology. I dont expect any user code (like loading a model to exist in the ingress deployment class)